### PR TITLE
template bug fixes

### PR DESCRIPTION
**Pull request by mjiang102628**
- add HTLC.get_transaction()
- new goldens
- includes changes from michelle/split
- new tests pass

### CODE REVIEW

File: algosdk/template.py [hunk 1]
```diff
@@ -174,6 +183,40 @@ def get_program(self):
         types = [int, int, "address", "base64", "address", int]
         return inject(orig, offsets, values, types)
 
+    @staticmethod
+    def get_transaction(contract, preimage, first_valid, last_valid, gh, fee):
+        """
+        Return a group transactions array which transfer funds according to
+        the contract's ratio.
```
Comment (winder) [inline algosdk/template.py]: Wrong comment `Return a transaction which will release funds if a matching preimage is used.`

File: algosdk/template.py [hunk 2]
```diff
@@ -174,6 +183,40 @@ def get_program(self):
...
+        if not (len(ints) == 4 and len(bytearrays) == 3):
+            raise error.WrongContractError("split")
+        max_fee = ints[0]
+        receiver = encoding.encode_address(bytearrays[0])
+
+        lsig = transaction.LogicSig(contract, [base64.b64decode(preimage)])
+        txn = transaction.PaymentTxn(logic.address(contract), fee, first_valid, last_valid, gh, None, 0, close_remainder_to=receiver)
+
+        if txn.fee > max_fee:
+            raise error.TemplateInputError("the transaction fee should not be greater than " + str(max_fee))
```
Comment (winder) [inline algosdk/template.py]: I think we should be able to verify the hash image as well

File: test_unit.py [hunk 3]
```diff
@@ -1209,6 +1230,21 @@ def test_HTLC(self):
...
+            "gqRsc2lngqNhcmeRxCB/g7Flf/H8U7ktwYFIodZd/C1LH6PWdyhK3dIAEm2Qaa" +
+            "FsxJcBIAToBwEAwM8kJgMg5pqWHm8tX3rIZgeSZVK+mCNe0zNjyoiRi7nJOKkV" +
+            "tvkgf4OxZX/x/FO5LcGBSKHWXfwtSx+j1ncoSt3SABJtkGkg/ryguxRKWk6ntD" +
+            "ikaBrIDmyhBby2B/xWUyXJVpX2ohMxASIOMRAjEhAxBzIDEhAxCCQSEDEJKBIt" +
+            "ASkSEDEJKhIxAiUNEBEQo3R4boelY2xvc2XEIOaalh5vLV96yGYHkmVSvpgjXt" +
+            "MzY8qIkYu5yTipFbb5o2ZlZc0D6KJmdgGiZ2jEIH+DsWV/8fxTuS3BgUih1l38" +
+            "LUsfo9Z3KErd0gASbZBpomx2ZKNzbmTEIFNCP4JtpWHGzW5H9EJSwasC1THntU" +
+            "IiTJGurfnrDvCqpHR5cGWjcGF5")
+        ltxn = template.HTLC.get_transaction(
+            p, "f4OxZX/x/FO5LcGBSKHWXfwtSx+j1ncoSt3SABJtkGk=", 1,
```
Comment (winder) [inline test_unit.py]: I don't think this is a valid input, don't we need to pass in something which, once hashed, is equal to `f4OxZX/x/FO5LcGBSKHWXfwtSx+j1ncoSt3SABJtkGk=`?

File: algosdk/template.py [hunk 4]
```diff
@@ -174,6 +184,56 @@ def get_program(self):
...
+        
+        Returns:
+            LogicSigTransaction: transaction to claim algos from
+                contract account
+        """
+        _, ints, bytearrays = logic.read_program(contract)
+        if not (len(ints) == 4 and len(bytearrays) == 3):
+            raise error.WrongContractError("split")
+        max_fee = ints[0]
+        hash_function = contract[-15]
```
Comment (winder) [inline algosdk/template.py]: :+1:

File: algosdk/template.py [hunk 5]
```diff
@@ -60,65 +63,72 @@ def get_program(self):
...
-        if amount % ratd == 0:
-            amt_1 = amount // ratd * ratn
-            amt_2 = amount - amt_2
-        elif precise:
-            raise error.NotDivisibleError
-        else:
-            amt_1 = round(amount / ratd * ratn)
+        if amount % (rat_1 + rat_2) == 0:
+            amt_1 = amount // (rat_1 + rat_2) * rat_1
             amt_2 = amount - amt_1
```
Comment (winder) [inline algosdk/template.py]: This is how I computed the amounts:
```
        Double fraction = Double.valueOf(rat1) / Double.valueOf(rat1+rat2);
        int receiverOneAmount = Long.valueOf(Math.round(fraction * amount)).intValue();
        int receiverTwoAmount = Long.valueOf(Math.round((1.0 - fraction) * amount)).intValue();
```

Comment (mjiang102628) [reply algosdk/template.py]: I think these are equivalent since amount must be divisible by (rat_1 + rat_2) because I've simplified the fraction first.

Comment (winder) [reply algosdk/template.py]: I didn't realize `//` did an implicit floor. There might be cases where the number might need to round up after the divide though. I noticed in my code that for our test values `receiverTwoAccount` needed to round up while `receiverOneAmount` needed to round down.

**Pull request merged**